### PR TITLE
[librdkafka] Add new port

### PR DESCRIPTION
**Pull request by myd7349**
#5920

### CODE REVIEW

**Comment by Rastaban** [comment]
I see failures on Linux and UWP.  Are those expected?

**Comment by myd7349** [comment]
@Rastaban I haven't tested it yet. :-)
Maybe I will test it after the holiday.

**Comment by myd7349** [comment]
Known issues:
1. x86-windows
   >1>   Creating library E:/vcpkg/buildtrees/rdkafka/src/v1.0.0-5c63a98714/examples/build/Debug/rdkafka_example_cpp.lib and object E:/vcpkg/buildtrees/rdkafka/src/v1.0.0-5c63a98714/examples/build/Debug/rdkafka_example_cpp.exp
1>rdkafka_example.obj : error LNK2001: unresolved external symbol "public: static int const RdKafka::Topic::PARTITION_UA" (?PARTITION_UA@Topic@RdKafka@@2HB)
1>rdkafka_example.obj : error LNK2001: unresolved external symbol "public: static __int64 const RdKafka::Topic::OFFSET_BEGINNING" (?OFFSET_BEGINNING@Topic@RdKafka@@2_JB)
1>rdkafka_example.obj : error LNK2001: unresolved external symbol "public: static __int64 const RdKafka::Topic::OFFSET_END" (?OFFSET_END@Topic@RdKafka@@2_JB)
1>rdkafka_example.obj : error LNK2001: unresolved external symbol "public: static __int64 const RdKafka::Topic::OFFSET_STORED" (?OFFSET_STORED@Topic@RdKafka@@2_JB)
1>E:\vcpkg\buildtrees\rdkafka\src\v1.0.0-5c63a98714\examples\build\Debug\rdkafka_example_cpp.exe : fatal error LNK1120: 4 unresolved externals

**Comment by NancyLi1013** [comment]
Update the latest test results from the current CI system:

|x64-uwp   |master| test |notes
|----------|------|------|-----
|librdkafka|      |Fail  |New Port

|arm-uwp   |master| test |notes
|----------|------|------|-----
|librdkafka|      |Fail  |New Port

Please help confirm if they are expected. Thanks.

**Comment by myd7349** [comment]
>     1. LNK2001

This issue seems a bug in the upstream repo: https://github.com/edenhill/librdkafka/pull/2274

**Comment by myd7349** [comment]
> 
> 
> Known issues:
> 
>     1. x86-windows
>        > 1>   Creating library E:/vcpkg/buildtrees/rdkafka/src/v1.0.0-5c63a98714/examples/build/Debug/rdkafka_example_cpp.lib and object E:/vcpkg/buildtrees/rdkafka/src/v1.0.0-5c63a98714/examples/build/Debug/rdkafka_example_cpp.exp
>        > 1>rdkafka_example.obj : error LNK2001: unresolved external symbol "public: static int const RdKafka::Topic::PARTITION_UA" (?PARTITION_UA@Topic@RdKafka@@2HB)
>        > 1>rdkafka_example.obj : error LNK2001: unresolved external symbol "public: static __int64 const RdKafka::Topic::OFFSET_BEGINNING" (?OFFSET_BEGINNING@Topic@RdKafka@@2_JB)
>        > 1>rdkafka_example.obj : error LNK2001: unresolved external symbol "public: static __int64 const RdKafka::Topic::OFFSET_END" (?OFFSET_END@Topic@RdKafka@@2_JB)
>        > 1>rdkafka_example.obj : error LNK2001: unresolved external symbol "public: static __int64 const RdKafka::Topic::OFFSET_STORED" (?OFFSET_STORED@Topic@RdKafka@@2_JB)
>        > 1>E:\vcpkg\buildtrees\rdkafka\src\v1.0.0-5c63a98714\examples\build\Debug\rdkafka_example_cpp.exe : fatal error LNK1120: 4 unresolved externals

This issue seems a bug in the upstream repo: https://github.com/edenhill/librdkafka/pull/2274

**Comment by webfolderio** [comment]
> This seems a bug in the upstream repo

It's fixed

**Comment by myd7349** [comment]
I will give those features a test sooner or later.

**Comment by myd7349** [comment]
https://github.com/edenhill/librdkafka/pull/2307

**Comment by myd7349** [comment]
OK. I have tested all features with this little project https://github.com/myd7349/Ongoing-Study/tree/master/cpp/CMake/vcpkg/librdkafka_pr_2274. These triplets are tested:
* x86-windows
* x86-windows-static
* x64-windows
* x64-windows-static
* x64-linux
I have removed the `WIP` tag and any review is welcome.

**Comment by myd7349** [comment]
I have gave `UWP` support a try, but failed: https://github.com/edenhill/librdkafka/files/3152143/install-x86-uwp-dbg-out.log
Maybe we can add UWP support in the future.

**Comment by webfolderio** [comment]
> Maybe we can add UWP support in the future.

Yes, go without UWP.

**Comment by NancyLi1013** [comment]
@myd7349, thanks for your all efforts on this port. The port only failed on arm64-windows now except for x64-uwp and arm-uwp. Have you tested on arm64-windows?
The error log on arm64-window is as follows:
[arm64-windows.zip](https://github.com/microsoft/vcpkg/files/3155302/arm64-windows.zip)

**Comment by myd7349** [comment]
Hi @NancyLi1013 I didn't have ARM toolchain installed on my PC, so I am not able to fix it currently.

**Comment by JackBoosY** [comment]
Hi @myd7349, can you use this patch to solve the problem of building errors under ARM/ARM64?
[fix-builderror-in-arm.zip](https://github.com/microsoft/vcpkg/files/3156667/fix-builderror-in-arm.zip)

Thanks.

**Comment by myd7349** [comment]
Hi @JackBoosY Thanks!
I will give it a try after work.

**Comment by myd7349** [comment]
Hi! @JackBoosY I am not sure whether it will work out by defining `get/put_unaligned*` as `get/put_unaligned_direct` since I saw these commits in [```snappy_compact.h```](https://github.com/edenhill/librdkafka/blob/v1.0.0/src/snappy_compat.h#L67-L74).
I have create an issue here: https://github.com/edenhill/librdkafka/issues/2324

**Comment by myd7349** [comment]
Hi! Can any one tell me whether the ARM patch work or not? If it works, maybe I can create a PR in the upstream.

File: ports/librdkafka/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,91 @@
+include(vcpkg_common_functions)
+
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO edenhill/librdkafka
+    REF dc85ffc8068c18dfdf118666b4c78de1c7ac640b
+    SHA512 b21be356806125ab8bdb74de41eef2c56714cfd2003065be362d9fab7e177a07675e95a13ed6d7fbd3dc38502194000b0e378ff6c43c7d54a0234ada0df881eb
+    HEAD_REF master
+    PATCHES
+        fix-arm-msvc-build
```
Comment (JackBoosY) [inline ports/librdkafka/portfile.cmake]: This line should be `fix-arm-msvc-build.patch`.

**Review by JackBoosY — COMMENTED** [review]
When using the patch, we do not need to write the path, but we must use the file name + suffix.

File: ports/librdkafka/portfile.cmake (hunk 1 above)
Comment (myd7349) [reply ports/librdkafka/portfile.cmake]: @JackBoosY Thanks!

**Comment by myd7349** [comment]
Hi! @JackBoosY I have fixed the patch filename. Does it work for ARM/ARM64 now? Thanks!

**Comment by JackBoosY** [comment]
It seems that this port does not support uwp.
Except for uwp, other triplets are built successfully.

**Comment by myd7349** [comment]
@JackBoosY Thanks!
BTW, it seems that I can see the CI log now: 
https://github.com/microsoft/vcpkg/pull/6618
https://github.com/microsoft/vcpkg/pull/6628
https://dev.azure.com/vcpkg/public/_build?definitionId=10&view=buildsHistory

**Comment by grdowns** [comment]
Thanks for the new port, @myd7349!

**Pull request merged**